lib/chef/provider/git.rb

```diff
@@ -224,16 +229,24 @@ def fetch_updates
           git("fetch", new_resource.remote, "--tags", cwd: cwd)
           if new_resource.checkout_branch
             # check out to a local branch
-            git("branch", "-f", new_resource.checkout_branch, target_revision, cwd: cwd)
-            git("checkout", new_resource.checkout_branch, cwd: cwd)
+            if already_on_target_branch?
+              git("reset", "--hard", target_revision, cwd: cwd)
+            else
+              git("branch", "-f", new_resource.checkout_branch, target_revision, cwd: cwd)
+              git("checkout", new_resource.checkout_branch, cwd: cwd)
+            end
           elsif sha_hash?(new_resource.revision) || is_tag?
             # detached head
             git("reset", "--hard", target_revision, cwd: cwd)
           else
             # need a branch with a tracking branch
-            git("branch", "-f", new_resource.revision, target_revision, cwd: cwd)
-            git("checkout", new_resource.revision, cwd: cwd)
-            git("branch", "-u", "#{new_resource.remote}/#{new_resource.revision}", cwd: cwd)
+            if already_on_target_branch?
+              git("reset", "--hard", target_revision, cwd: cwd)
+            else
+              git("branch", "-f", new_resource.revision, target_revision, cwd: cwd)
+              git("checkout", new_resource.revision, cwd: cwd)
+              git("branch", "-u", "#{new_resource.remote}/#{new_resource.revision}", cwd: cwd)
+            end
           end
```

Comment:
I'm not very familiar with how this code is supposed to work, but I'm curious if some of the branches of this conditional could be combined to reduce the nested conditions. If the check for `if new_resource.checkout_branch` does not strictly need to come first, you could do something like this, for example:

```
if sha_hash?(new_resource.revision) || is_tag? || already_on_target_branch?
   git("reset", "--hard", target_revision, cwd: cwd)
elsif new_resource.checkout_branch
  # ...
else
  # ...
end
```

Reply:
I think the logic would work, but not sure if the code wouldn't be harder to read

Reply:
I pushed your change.